### ISSUE TITLE
Support stream response in browser

### ISSUE THREAD

**Issue by olalonde** [mzabriskie/axios#479]
Would be useful if `responseType` could be set to `stream` in the browser. Right now, it only works in node.

**Comment by nickuraltsev** [comment]
Are you suggesting to add support for [this](https://streams.spec.whatwg.org/)? Another reason to create a `fetch` adapter (#484) as it's not possible to implement this using `XHR`.

**Comment by olalonde** [comment]
I was more hoping for a Node.js readable stream but I guess a whatwg stream would be better than nothing :)

**Comment by nickuraltsev** [comment]
`XMLHttpRequest` does not support response body streaming, so I don't really understand the point of wrapping an `XMLHttpRequest` response into a stream. (`fetch` standard does support streaming so adding support for `fetch` response streams makes sense to me.) 

Could you please elaborate on the problem you are solving using `xhr-stream`?

**Comment by olalonde** [comment]
I have a long lived http response that emits new line delimited JSON objects, and want to start display the result in the browser without waiting for the response to finish. `xhr-stream` seems to work although I'm not sure how it works. I am piping the xhr stream to https://github.com/jahewson/node-byline which is why a Node.js compatible read stream would have been helpful. I'm sure there are probably whatwg stream to node.js stream wrappers somewhere though.

**Comment by mividtim** [comment]
Why was this closed?  Actually, XHR _does_ support chunked reading, through readyState 3 rather than 4.

**Comment by mividtim** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState

**Comment by mividtim** [comment]
You closed 505 as a duplicate of this issue, and then closed this issue.  Neither has been resolved.  This issue should be re-opened.

**Comment by amazingandyyy** [comment]
omg both #505 and #479 are closed now wt....

**Comment by rizkyramadhan** [comment]
any news about this ?

**Comment by cgomezmendez** [comment]
What about react-native?

**Comment by affanshahid** [comment]
Any news on this?

**Comment by gwh-cpnet** [comment]
Would anybody provide a walkaround for it?

**Comment by jcjolley** [comment]
@gwh-cpnet
I'm stuck in IE11, so I'm unable to use the fetch API.  I have a work around, but it tailored to my use case of passing a large sequence of JSON objects from the database to an Office-js Excel Add-in.  The promise is only there so I can know when it finishes.  The callback does the actual work of handling the input.


`export const ie11XHRStreamHandler = async (url: string, callback: any) => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url, true);
  xhr.responseType = 'stream' as any; // IE11 specific. This prevents msCaching
  let seenChars = 0;
  const p = new Promise((resolve, reject) => {
    xhr.onreadystatechange = () => {
      if (xhr.readyState === xhr.LOADING) {
        callback(xhr.response.substr(seenChars));
        seenChars = xhr.responseText.length;
      } else if (xhr.readyState === xhr.DONE) {
        if (xhr.status === 200) {
          resolve();
        } else {
          reject(`XHR Failed with status ${xhr.status}: ${xhr.statusText}`);
        }
      }
    };
  });
  xhr.send();
  return p;
}`

**Comment by vizcay** [comment]
Ended up using https://github.com/eBay/jsonpipe, maybe someone find this useful.

**Comment by xintaoLi** [comment]
try this: 
1、set responseType: 'blob'
2、read response as text:
  response.text()  or read as response.stream() or response.arrayBuffer()

3、parse text to json or replace with regular

```
axios.post('url', params, {
                        cancelToken: this.source.token,
                        responseType: 'blob'
                    })
                    .then(response => {
                        response.text().then(text => {
                            const reg = /:\s*(\d{15,25})(\s*\}|,?)/
                            const textdata = text.replace(reg, ':"$1"$2')
                            let res = JSON.parse(textdata)
     })
})
```

**Comment by leohxj** [comment]
So what is the right type of response data when responseType set 'stream'?

**Comment by shresthapradip** [comment]
Any update on this?

**Comment by dlrandy** [comment]
`export const getStreamFile = (url) => {
    return axios
        .get(url, {
            responseType: 'blob',
        })
        .then((res) => {
            const { data, headers } = res;
            const fileName = headers['content-disposition'].replace(
                /\w+;filename=(.*)/,
                '$1'
            );
            const blob = new Blob([data], { type: headers['content-type'] });
            const dom = document.createElement('a');
            const href = window.URL.createObjectURL(blob);
            dom.href = href;
            dom.download = decodeURI(fileName);
            dom.style.display = 'none';
            document.body.appendChild(dom);
            dom.click();
            dom.parentNode.removeChild(dom);
            window.URL.revokeObjectURL(href);
        });
};`

**Comment by 1isten** [comment]
Using `blob` as a workaround is meaningless as the entire response data is already loaded into memory.

The advantage of `stream` is that we can load the response chunk by chunk (hence reduce memory usage) and we can start to process the response as soon as possible.

**Comment by wufeiwua** [comment]
Any news on this? I have tried setting `responseType` to `stream`, but it doesn't work

**Comment by Cheaterman** [comment]
Bump :-)

**Comment by atonamy** [comment]
Such a disappointment and shame that in 2023 is still not supported even when already large language models heavily using chunked reading in their API.

**Comment by dante01yoon** [comment]
wow, I'm surprised  axios still doesn't support stream responseType on client. 
It's huge pain point adapt fetch api in only few rest apis in manner of deal with error response and middleware

**Comment by fengerzh** [comment]
> wow, I'm surprised axios still doesn't support stream responseType on client. It's huge pain point adapt fetch api in only few rest apis in manner of deal with error response and middleware

Even `fetch` is not perfect, as it does not support `onmessage` event, you cannot see messages in network console. The best way is to combine 2 requests, one use fetch to POST content, and the other use `onmessage` event of `new EventSource` to get messages as it does not support POST request.

**Comment by blackbing** [comment]
I can't believe axios doesn't support fetch stream in browser.

**Comment by blackbing** [comment]
@fengerzh fetch function can achieve this. here is an example

https://gist.github.com/blackbing/22ed6db703727fb050a071ce2911684f#file-fetchstream-js

**Comment by mividtim** [comment]
There is an issue with fetch, however... It doesn't support streaming uploads.

**Comment by jetjokers24895** [comment]
I got the same issue when using Axios with Electron. I found a solution to this. With Axios 1.4.0. I config adapter

axios({
URL: 'abc.com',
...
adapter: 'http'
})

**Comment by 3rd** [comment]
> I got the same issue when using Axios with Electron. I found a solution to this. With Axios 1.4.0. I config adapter
> 
> axios({ URL: 'abc.com', ... adapter: 'http' })

The http adapter is not available in browsers.

**Comment by lukelu520** [comment]
The issue should be reopened since axios still doesn't support stream response. Correct me if I am wrong, hope to hear your expertise.

**Comment by emargin** [comment]
BUMP .-.

**Comment by AkdM** [comment]
I couldn't make the `responseType` stream option work too, but I made a workaround with the `onDownloadProgress` option, like such:

```js
axios({
  // ...options,
  onDownloadProgress: (evt) => {
    // Parse response from evt.event.target.responseText || evt.event.target.response
    // The target holds the accumulator + the current response, so basically everything from the beginning on each response
  }
})
```

**Comment by psankar** [comment]
>     // Parse response from evt.event.target.responseText || evt.event.target.response
>     // The target holds the accumulator + the current response, so basically everything from the beginning on each response

Thanks for this @AkdM . But do you know of a way to get only the new responses instead of the accumulated response since the beginning ?

**Comment by jonah-butler** [comment]
> I couldn't make the `responseType` stream option work too, but I made a workaround with the `onDownloadProgress` option, like such:
> 
> ```js
> axios({
>   // ...options,
>   onDownloadProgress: (evt) => {
>     // Parse response from evt.event.target.responseText || evt.event.target.response
>     // The target holds the accumulator + the current response, so basically everything from the beginning on each response
>     // Note that it's evt.target instead of evt.event.target for older axios versions
>   }
> })
> ```
This is the same approach I took as well for collecting chunked responses from a stream. Though the **evt** param in the onDownloadProgress callback was creating some issues for me in TS.  It seems the type that was added is declared as **[ProgressEvent](https://github.com/axios/axios/pull/4675/commits/c3ed3198498e501390269eb97a837b1a5a79764d)**. With that being the case, I had to cast evt.target(or evt.event.target) to an XMLHttpRequest before I could access the **.response** property, otherwise I get a _Property 'response' does not exist on type 'EventTarget'._ error.

```typescript
onProgress(e: ProgressEvent): void {
      const req = e.target as XMLHttpRequest;
      this.textChunk = req.response;
}

async function streamedResponseHandler(
  cb: (e: ProgressEvent) => void
): Promise<void> {
  axios.get(
    "/endpoint",
    {
      onDownloadProgress: cb,
    }
  );
}
```
This callback approach was good for my use case in Vue so I could retain the reactivity of `this.textChunk`, which is a data property, so as this.textChunk was getting reassigned from `req.response`, all DOM updates would just happen.



Then just calling it:
```typescript
streamedResponseHandler(onProgress);
```

**Comment by AkdM** [comment]
Interesting, thanks. I am going to make a transition to TS very soon so that will help me.

**Comment by jonah-butler** [comment]
Thanks for your example too @AkdM. That was very helpful for me.

**Comment by ZulluBalti** [comment]
> > ```
> > // Parse response from evt.event.target.responseText || evt.event.target.response
> > // The target holds the accumulator + the current response, so basically everything from the beginning on each response
> > ```
> 
> Thanks for this @AkdM . But do you know of a way to get only the new responses instead of the accumulated response since the beginning ?

Hi, did you find any build-in elegant solution?

**Comment by jonah-butler** [comment]
@ZulluBalti I know this isn't a built-in, but in my example based on @AkdM, slicing the response at the previous length of `this.textChunk` is a straightforward way of capturing only the new data without modifying the response text.
``` typescript
onProgress(e: ProgressEvent): void {
      const req = e.target as XMLHttpRequest;
      console.log(req.response.slice(this.scriptChunk.length); // <---- only new data from each new response
      this.textChunk = req.response;
}
```

**Comment by ZulluBalti** [comment]
> I couldn't make the `responseType` stream option work too, but I made a workaround with the `onDownloadProgress` option, like such:
> 
> ```js
> axios({
>   // ...options,
>   onDownloadProgress: (evt) => {
>     // Parse response from evt.event.target.responseText || evt.event.target.response
>     // The target holds the accumulator + the current response, so basically everything from the beginning on each response
>     // Note that it's evt.target instead of evt.event.target for older axios versions
>   }
> })
> ```

Hi, I'm facing one problem with this approach.

Axios buffers the chunks and return multiple chunks.
Is there any way to run this function every time the server writes to the stream?

for example, in the backend I've this code

```const chunks = ["Chunk 1 ", "Chunk 2 ", "Chunk 3"];
for  (const chunk of chunks) {
  res.write(chunk);
  await sleep(1000); // custom function to wait for 1 second
}
res.end```

I was expecting to get the chunks like this

```"Chunk 1"``` when ```onDownloadProgress``` runs for the first time
```Chunk1 Chunk2``` when it runs for the 2nd time
```Chunk1 Chunk2 Chunk3``` when it runs for the 3rd time

But I'm getting it like this
```Chunk 1Chunk 2``` (multiple chunks at a time)

**Comment by brunolm** [comment]
> as it's not possible to implement this using `XHR`.

I guess this reply is over 9 years old. ^

XHR + Streaming text:

```js
var xhr = new XMLHttpRequest();
xhr.open('GET', '/endpoint', true);
xhr.timeout = 10000;
xhr.ontimeout = function() {
  console.error('Request timed out!');
};
xhr.onprogress = function() {
  var responseText = xhr.responseText;
  var chunk = responseText.slice(xhr.prevLen);
  xhr.prevLen = responseText.length;
  console.log(chunk);
};
xhr.onload = function() {
  console.log('Done!');
};
xhr.send();
```

Next.js endpoint:

```ts
import { NextApiResponse } from 'next'

// https://developer.mozilla.org/docs/Web/API/ReadableStream#convert_async_iterator_to_stream
function iteratorToStream(iterator: any) {
  return new ReadableStream({
    async pull(controller) {
      const { value, done } = await iterator.next()

      if (done) {
        controller.close()
      } else {
        controller.enqueue(value)
      }
    },
  })
}

function sleep(time: number) {
  return new Promise((resolve) => {
    setTimeout(resolve, time)
  })
}

const encoder = new TextEncoder()

async function* makeIterator() {
  yield encoder.encode('<p>One</p>')
  await sleep(700)
  yield encoder.encode('<p>Two</p>')
  await sleep(700)
  yield encoder.encode('<p>Three</p>')
}

export default async function handler(req, res: NextApiResponse<any>) {
  console.log('-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=')
  console.log('makeIterator')
  const iterator = makeIterator()
  console.log('iteratorToStream')
  const stream = iteratorToStream(iterator)

  console.log('-=-=- END =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=')

  return new Response(stream)
}

export const runtime = 'edge'
```

**Comment by leookun** [comment]
why close

**Comment by Cheaterman** [comment]
Because the author has no intention to implement this feature on current Axios. Meanwhile #4209 is closed and #4477 is locked - and both are stale, last activity in 2022. If anyone still cares about this, I recommend you use `fetch` instead.

**Comment by mividtim** [comment]
Would the maintainer kindly hand maintenance over to someone willing to actually maintain the package? Fetch is garbage. The API is inferior, and it doesn't support streaming uploads at all.

**Comment by leookun** [comment]
> Because the author has no intention to implement this feature on current Axios. Meanwhile #4209 is closed and #4477 is locked - and both are stale, last activity in 2022. If anyone still cares about this, I recommend you use `fetch` instead.

fetch solved my problem，I implemented  axios-like function with fetch , maybe I should use fetch from the beginning

**Comment by Thorvarium** [comment]
this should be reopened LOL

**Comment by Thorvarium** [comment]
@brunolm I don't think that the fetch version works. It is waiting for the entire response before starting to print

**Comment by brunolm** [comment]
> @brunolm I don't think that the fetch version works. It is waiting for the entire response before starting to print

Did you set this?

```
export const runtime = 'edge'
```

Maybe it's your next version or whatever backend you're using that needs some config.

**Comment by gofish543** [comment]
> This should be reopened LOL. Shocking.

Very shocking.

**Comment by oneyoung19** [comment]
> > as it's not possible to implement this using `XHR`.
> 
> I guess this reply is over 9 years old. ^
> 
> XHR + Streaming text:
> 
> ```js
> var xhr = new XMLHttpRequest();
> xhr.open('GET', '/endpoint', true);
> xhr.timeout = 10000;
> xhr.ontimeout = function() {
>   console.error('Request timed out!');
> };
> xhr.onprogress = function() {
>   var responseText = xhr.responseText;
>   var chunk = responseText.slice(xhr.prevLen);
>   xhr.prevLen = responseText.length;
>   console.log(chunk);
> };
> xhr.onload = function() {
>   console.log('Done!');
> };
> xhr.send();
> ```
> 
> Bonus: Using fetch
> 
> ```ts
> const response = await fetch("/endpoint", {
>   signal: AbortSignal.timeout(10000),
> });
> 
> const reader = response.body.getReader();
> 
> while (true) {
>   const { done, value } = await reader.read();
> 
>   if (done) {
>     console.log("Stream complete");
>     break;
>   }
> 
>   console.log(new TextDecoder().decode(value));
> }
> ```
> 
> Axios:
> 
> ```
> Not supported. Said to be impossible in 2009, closed in 2016, and there's still no support for it today.
> ```
> 
> Next.js 14 endpoint:
> 
> ```ts
> import { NextApiResponse } from 'next'
> 
> // https://developer.mozilla.org/docs/Web/API/ReadableStream#convert_async_iterator_to_stream
> function iteratorToStream(iterator: any) {
>   return new ReadableStream({
>     async pull(controller) {
>       const { value, done } = await iterator.next()
> 
>       if (done) {
>         controller.close()
>       } else {
>         controller.enqueue(value)
>       }
>     },
>   })
> }
> 
> function sleep(time: number) {
>   return new Promise((resolve) => {
>     setTimeout(resolve, time)
>   })
> }
> 
> const encoder = new TextEncoder()
> 
> async function* makeIterator() {
>   yield encoder.encode('<p>One</p>')
>   await sleep(700)
>   yield encoder.encode('<p>Two</p>')
>   await sleep(700)
>   yield encoder.encode('<p>Three</p>')
> }
> 
> export default async function handler(req, res: NextApiResponse<any>) {
>   const iterator = makeIterator()
>   const stream = iteratorToStream(iterator)
> 
>   return new Response(stream)
> }
> 
> export const runtime = 'edge'
> ```

The `chatgpt` may used `XHR + streaming`.
<img width="1669" alt="image" src="https://github.com/axios/axios/assets/37331643/bbba0283-fae0-43ee-9747-109956e28d9e">

**Comment by mankeheaven** [comment]
use fetch adapter

![image](https://github.com/axios/axios/assets/17426397/391f8c9d-b8fb-4b82-9ede-a78c94d3b70e)

**Comment by lagoasoft-yagosantos** [comment]
> > as it's not possible to implement this using `XHR`.
> 
> I guess this reply is over 9 years old. ^
> 
> XHR + Streaming text:
> 
> ```js
> var xhr = new XMLHttpRequest();
> xhr.open('GET', '/endpoint', true);
> xhr.timeout = 10000;
> xhr.ontimeout = function() {
>   console.error('Request timed out!');
> };
> xhr.onprogress = function() {
>   var responseText = xhr.responseText;
>   var chunk = responseText.slice(xhr.prevLen);
>   xhr.prevLen = responseText.length;
>   console.log(chunk);
> };
> xhr.onload = function() {
>   console.log('Done!');
> };
> xhr.send();
> ```

Since I do not see anyone talking about it, this solution works as well on react-native, searching about hout to handle text stream requests for native I can't find another solution besides the solution on using fetch API polyfill, which doesn't work well for me, this XMLHttpRequest was the best solution for me, works like a charm, I just needed to make an addition for my case adding ` xhr.setRequestHeader('Content-Type', 'application/json');`, but I don't know if this is related to my backend architecture, anyways thanks @brunolm

**Comment by marcinfigiel** [comment]
Wow, this caught me off-guard. Seems like a good reason to ditch axios :(

**Comment by marcinfigiel** [comment]
> @marcinfigiel, we have the fetch adaptor to assist in making this possible unless I misunderstand what's needed.

Right, that works perfectly fine, thanks!
